README.md
# IdeaSpace 360 theme
This theme comes pre-installed with the <a href="https://github.com/IdeaSpaceVR/IdeaSpace">IdeaSpaceVR CMS</a>.

![IdeaSpace-360](screenshot.jpg)

More info: <a href="https://www.ideaspacevr.org/themes/ideaspace-360">https://www.ideaspacevr.org/themes/ideaspace-360</a>.


CHANGELOG.md
## 2.0

- Updated to A-Frame v0.9.2
- Removed animation component (deprecated)
- Added anti-aliasing

## 1.7

- Hide controller models if HMD is connected but site is used in desktop mode
- Updated to A-Frame 0.8.2

## 1.6

- Deprecated #theme-compatibility config parameter
- Fixed 6-DOF controller support

## 1.5

- user height and positional controllers height fix
- photo sphere menu bug fix: clicking on thumbnail of current photo sphere made scene black
- prints additional Chrome origin trial token infos in meta tag

## 1.4

- Controller responsiveness improved on navigation and info hotspots
- Version bump to A-Frame v0.7.0

## 1.3

- Better desktop and mobile support, without head-mounted display
- Do not show cursor if no head-mounted display is connected
- Changed hotspot style

## 1.2

- Select style and color for sphere titles and hotspot text annotations 
- Support for Samsung Gear VR / Daydream / Oculus Rift / HTC Vive; controller required
- Desktop and mobile / touch screen support


js/isvr-scene.js
AFRAME.registerComponent('isvr-scene', {

  
    init: function () {

        if (!AFRAME.utils.device.checkHeadsetConnected()) {
            document.querySelector('#no-hmd-intro').setAttribute('visible', true);
            //document.querySelector('#cursor').setAttribute('geometry', { radius: 0.04 });
        }

        this.el.addEventListener('enter-vr', function() {

						if (AFRAME.utils.device.checkHeadsetConnected()) {
                document.querySelector('#camera-wrapper').setAttribute('position', {x: 0, y: 0, z: 0});
            }

            /* workaround for non-positional tracking devices and a-frame 0.7.0 */
            /*if (!AFRAME.utils.device.checkHasPositionalTracking()) {
                document.querySelector('#camera').setAttribute('camera', {
                    far: 10000,
                    fov: 80,
                    near: 0.1
                });
            }*/

						/* show controllers only in VR */
            var laser_controls = document.querySelectorAll('.laser-controls');
            for (var i = 0; i < laser_controls.length; i++) {
                laser_controls[i].setAttribute('visible', true);
            }

            document.querySelector('a-scene').addState('entered-vr');

        });

        this.el.addEventListener('exit-vr', function() {

            location.reload();

        });

        var laser_controls = document.querySelectorAll('.laser-controls');

        for (var i = 0; i < laser_controls.length; i++) {

            laser_controls[i].addEventListener('controllerconnected', function (evt) {
                document.querySelector('a-scene').addState('controllerconnected');
                document.querySelector('a-scene').setAttribute('data-controllertype', evt.detail.name);
                //document.querySelector('#cursor').setAttribute('visible', false);
            });

        }

    }

});



js/isvr-init-assets-component.js
AFRAME.registerComponent('isvr-init-assets', {

    schema: {
        url: {
            default: ''
        }
    },

    init: function() {

        this.xmlhttp = new XMLHttpRequest();
        this.xmlhttp.onreadystatechange = this.responseHandler.bind(this);
        this.xmlhttp.open('GET', this.data.url, true);
        this.xmlhttp.send();
    },

    responseHandler: function() {

        var self = this;

        if (this.xmlhttp.readyState == 4 && this.xmlhttp.status == 200) {

            var obj = JSON.parse(this.xmlhttp.responseText);

            /* 1st photo sphere asset, to show immediately */
            var photosphere_image = new Image();
            photosphere_image.onload = (function(content_id) {
                return function() {
                    var image_elem = document.createElement('img');
                    image_elem.setAttribute('id', 'img-photosphere-1');
                    image_elem.setAttribute('data-content-id', content_id);
                    image_elem.setAttribute('src', this.src);

                    var assets = document.querySelector('a-assets');
                    assets.appendChild(image_elem);

                    var sphere = document.querySelector('#photosphere');
                    sphere.setAttribute('material', 'src', '#img-photosphere-1');
                    sphere.setAttribute('data-content-id', content_id);
                    //sphere.emit('photosphere-fade-in');

                    var photosphere_texture_loaded_listener = function() {

                        document.querySelector('#photosphere-loading').setAttribute('visible', false);
                        document.querySelector('#photosphere-loading-background').setAttribute('visible', false);
                        document.querySelector('#photosphere-loading').emit('stop-photosphere-loading-anim');

                        document.querySelector('#photosphere-start-btn').setAttribute('visible', true);

                        document.querySelector('#intro-0').addEventListener('click', function(evt) {

														document.querySelector('#photosphere-start-btn-circle').emit('stop-photosphere-start-btn-anim');

                            // workaround because of interference with menu
                            document.querySelector('#intro-0').setAttribute('position', { x: 0, y: 1.6, z: -10 });
                            document.querySelector('#no-hmd-intro').setAttribute('position', { x: 0, y: 1.6, z: -10 });

                            document.querySelector('#intro-0').setAttribute('visible', false);
                            document.querySelector('#no-hmd-intro').setAttribute('visible', false);

                            var hotspots = document.querySelectorAll('.hotspot-wrapper-content-id-' + content_id);
                            for (var i = 0; i < hotspots.length; i++) {
                                hotspots[i].setAttribute('visible', true);
                            }
                            hotspots = document.querySelectorAll('.hotspot-content-id-' + content_id);
                            for (var i = 0; i < hotspots.length; i++) {
                                hotspots[i].setAttribute('visible', true);
                            }

                            var title = document.querySelector('#photosphere-title-content-id-' + content_id);
                            if (title != null && title.getAttribute('data-shown') == 'false') {
                                title.setAttribute('position', { x: 0, y: 1.6, z: -2.1 });
                                title.setAttribute('visible', true);
                                title.setAttribute('data-shown', 'true');
                                setTimeout(function() {
                                    title.setAttribute('visible', false);
                                    // workaround because of interference with menu 
                                    title.setAttribute('position', { x: 0, y: 10, z: -2.1 });
                                }, 10000);
                            }
                            sphere.removeEventListener('materialtextureloaded', photosphere_texture_loaded_listener);
                        });

                    };
                    sphere.addEventListener('materialtextureloaded', photosphere_texture_loaded_listener); 
                }
            }(obj['photo-spheres'][0]['photo-sphere']['#content-id']));
            photosphere_image.src = obj['photo-spheres'][0]['photo-sphere']['#uri']['#value'];


            /* photo sphere thumbnail assets */
            for (var i=0; i<obj['photo-spheres'].length; i++) {

                var id = obj.from + i;

                var photosphere_thumb_image = new Image();
                photosphere_thumb_image.onload = (function(id, i, content_id) {
                    return function() {
                        var image_elem = document.createElement('img');
                        image_elem.setAttribute('id', 'img-photosphere-thumb-' + id);
                        image_elem.setAttribute('data-content-id', content_id);
                        image_elem.setAttribute('src', this.src);
                        var assets = document.querySelector('a-assets');
                        assets.appendChild(image_elem);

                        var thumb = document.querySelector('#photosphere-thumb-' + (i+1));

                        // img-photosphere-1 has already been loaded 
                        if (id != 1) {

                            var photosphere_image = new Image();
                            photosphere_image.onload = (function(id, thumb, content_id) {
                                return function() {
                                    var image_elem = document.createElement('img');
                                    image_elem.setAttribute('id', 'img-photosphere-' + id);
                                    image_elem.setAttribute('data-content-id', content_id);
                                    image_elem.setAttribute('src', this.src);
                                    var assets = document.querySelector('a-assets');
                                    assets.appendChild(image_elem);

                                    document.querySelector('#photosphere-loading-' + (i+1)).setAttribute('visible', false);
                                    document.querySelector('#photosphere-loading-' + (i+1)).emit('stop-photosphere-loading-anim-' + (i+1));
                        
                                    thumb.setAttribute('material', 'src', '#img-photosphere-thumb-' + id);
                                    thumb.setAttribute('visible', true);
                                }
                            }(id, thumb, content_id));
                            photosphere_image.src = obj['photo-spheres'][i]['photo-sphere']['#uri']['#value'];

                        } else {

                          document.querySelector('#photosphere-loading-' + (i+1)).setAttribute('visible', false);
                          document.querySelector('#photosphere-loading-' + (i+1)).emit('stop-photosphere-loading-anim-' + (i+1));

                          thumb.setAttribute('material', 'src', '#img-photosphere-thumb-' + id);
                          thumb.setAttribute('visible', true);
                        } 
                    }
                }(id, i, obj['photo-spheres'][i]['photo-sphere']['#content-id']));
                photosphere_thumb_image.src = obj['photo-spheres'][i]['photo-sphere']['photo-sphere-navigation-preview-image']['#uri']['#value'];

            } /* for */ 


        } /* if */

    }

});




js/isvr-photosphere-menu-component.js
AFRAME.registerComponent('isvr-photosphere-menu', {

    /* in play: otherwide AFRAME.utils.device.checkHeadsetConnected() returns false */
    play: function() {

        if (AFRAME.utils.device.checkHeadsetConnected()) {

            document.querySelector('#photosphere').addEventListener('click', this.onClick.bind(this));

        } else if (!AFRAME.utils.device.checkHeadsetConnected() && AFRAME.utils.device.isMobile()) {

            document.querySelector('#photosphere').addEventListener('click', this.onClick.bind(this));

        } else {

            window.onkeydown = (function(e) {
                var code = e.keyCode ? e.keyCode : e.which;
                /* space or enter keys */
                if (code === 13 || code === 32) { 
                    this.onClick();
                }
            }).bind(this);
        }


        this.yaxis = new THREE.Vector3(0, 1, 0);
        this.zaxis = new THREE.Vector3(0, 0, 1);

        this.pivot = new THREE.Object3D();
        this.el.object3D.position.set(0, 1.6, -4);

        this.el.sceneEl.object3D.add(this.pivot);
        this.pivot.add(this.el.object3D);

    },

    onClick: function(evt) {

        var hide_hotspot_text = false;
        var hotspot_text = document.querySelectorAll('.hotspot-text');
        for (var i = 0; i < hotspot_text.length; i++) {
            if (hotspot_text[i].getAttribute('visible') == true) {
                hotspot_text[i].setAttribute('visible', false);
                /* hide hotspot texts on this click first, next click is for showing menu */
                hide_hotspot_text = true;
            }
        }            

        var hide_photosphere_title = false;
        var photosphere_title = document.querySelectorAll('.photosphere-title');
        for (var i = 0; i < photosphere_title.length; i++) {
            if (photosphere_title[i].getAttribute('visible') == true) {
                photosphere_title[i].setAttribute('visible', false);
                /* hide photo sphere title on this click first, next click is for showing menu */
                hide_photosphere_title = true;
            }
        }

        var content_id = document.querySelector('#photosphere').getAttribute('data-content-id');
        var hotspots = document.querySelectorAll('.hotspot-wrapper-content-id-' + content_id);
        for (var i = 0; i < hotspots.length; i++) {
            hotspots[i].setAttribute('visible', true);
        }



        if (hide_photosphere_title == false && 
						hide_hotspot_text == false && 
						document.getElementsByClassName('img-photosphere-thumb').length > 1 && 
            this.el.getAttribute('visible') == false && 
            document.querySelector('#photosphere-loading').getAttribute('visible') == false) {

            var hotspot_wrapper = document.querySelectorAll('.hotspot');
            for (var i = 0; i < hotspot_wrapper.length; i++) {
                hotspot_wrapper[i].setAttribute('visible', false);
            }            

            var direction = this.zaxis.clone();
            direction.applyQuaternion(document.querySelector('#camera').object3D.quaternion);
            var ycomponent = this.yaxis.clone().multiplyScalar(direction.dot(this.yaxis));
            direction.sub(ycomponent);
            direction.normalize();

            this.pivot.quaternion.setFromUnitVectors(this.zaxis, direction);

            this.el.setAttribute('position', { x: 0, y: 1.6, z: -4 });
            this.el.setAttribute('visible', true);

        } else if (hide_photosphere_title == false && 
						hide_hotspot_text == false && 
						document.getElementsByClassName('img-photosphere-thumb').length > 1 && 
						this.el.getAttribute('visible') == true) {

            this.el.setAttribute('position', { x: 0, y: 10, z: -4 });
            this.el.setAttribute('visible', false);

            var content_id = document.querySelector('#photosphere').getAttribute('data-content-id');

            var hotspots = document.querySelectorAll('.hotspot-content-id-' + content_id);
            for (var i = 0; i < hotspots.length; i++) {
                hotspots[i].setAttribute('visible', true);
            }

        }

    },

    update: function(oldData) {},

    remove: function() {}

});




js/isvr-photosphere-menu-thumb-component.js
AFRAME.registerComponent('isvr-photosphere-menu-thumb', {

		schema: {
        id: {
            default: ''
        }
    },


    init: function() {

    		this.el.addEventListener('click', this.onClick.bind(this));
      	this.el.addEventListener('mouseenter', this.onMouseenter.bind(this));
      	this.el.addEventListener('mouseleave', this.onMouseleave.bind(this));

				window.menu_thumb_enter = false;
    },


    onMouseenter: function(evt) {
				if (window.menu_thumb_enter == false) {
						document.querySelector(this.data.id).emit('trigger-mouseenter');
						window.menu_thumb_enter = true;
				}
		},


    onMouseleave: function(evt) {
				if (window.menu_thumb_enter == true) {
						document.querySelector(this.data.id).emit('trigger-mouseleave');
						window.menu_thumb_enter = false;
				}
		},


    onClick: function(evt) {

        var self = this;
				var el = document.querySelector(this.data.id);

        var position = el.getAttribute('position');

        /* prevent immediate selection of image after menu appears */
        if (el.parentEl.getAttribute('visible') && position.z == 0.5) {

            var image_id = el.getAttribute('data-image-id');
            var content_id = el.getAttribute('data-content-id');

            var hotspots = document.querySelectorAll('.hotspot-wrapper');
            for (var i = 0; i < hotspots.length; i++) {
                hotspots[i].setAttribute('visible', false);
            }

            var sphere = document.querySelector('#photosphere');

            /* do not allow selecting the current photo sphere again */
            if (sphere.getAttribute('material').src.id != 'img-photosphere-' + image_id) {

                sphere.emit('photosphere-fade-out');
                sphere.setAttribute('material', 'src', '#img-photosphere-' + image_id);
                sphere.setAttribute('data-content-id', content_id);

                var materialtextureloaded_listener = function() {

                    /* reset camera rotation in order to let people see the photo sphere title, if set */
                    var camera = document.querySelector('#camera');
                    camera.setAttribute('rotation', { x:0, y:0, z:0 });

                    document.querySelector('#photosphere-menu').setAttribute('visible', false);
                    document.querySelector('#photosphere-menu').setAttribute('position', { x: 0, y:10, z:-4 });

                    /* set visible to true on hotspot wrapper, opacity is still 0 so they are invisible */
                    var hotspot_wrapper = document.querySelectorAll('.hotspot-wrapper-content-id-' + content_id);
                    for (var i = 0; i < hotspot_wrapper.length; i++) {
                        hotspot_wrapper[i].setAttribute('visible', true);
                    }
                    /* animation */
                    var hotspots = document.querySelectorAll('.hotspot-content-id-' + content_id);
                    for (var i = 0; i < hotspots.length; i++) {
                        hotspots[i].setAttribute('visible', true);
                        hotspots[i].emit('hotspot-intro-' + content_id);
                    }

                    var title = document.querySelector('#photosphere-title-content-id-' + content_id);
                    if (title != null) {
                        title.setAttribute('position', { x: 0, y:1.6, z:-2 });
                        title.setAttribute('visible', true);
                        setTimeout(function() {
                            title.setAttribute('visible', false);
                            /* workaround because of interference with menu */
                            title.setAttribute('position', { x: 0, y:1.6, z:-10 });
                        }, 10000);
                    }
                    sphere.emit('photosphere-fade-in');
                    sphere.removeEventListener('materialtextureloaded', materialtextureloaded_listener);
                };
                sphere.addEventListener('materialtextureloaded', materialtextureloaded_listener);

            } /* if */

        } /* if */

    },

    update: function(oldData) {},

    remove: function() {}

});


views/scene.blade.php
@extends('theme::index')

@section('title', $space_title)

@section('scene')

    <a-scene isvr-scene renderer="antialias: true" background="color: #000000" loading-screen="dotsColor: #FFFFFF; backgroundColor: #000000">

    @include('theme::assets')


				<!--a-entity log geometry="primitive: plane" material="color: #111" text="color: lightgreen" position="0 1.6 -4"></a-entity//-->


        <a-entity id="camera-wrapper" position="0 1.6 0">
            <a-entity
                id="camera" 
                camera
                look-controls>
                <a-entity
                    cursor="fuse: false; rayOrigin: mouse"
                    id="cursor"
                    raycaster="objects: .collidable; far:5000"
                    position="0 0 -1.9"
                    geometry="primitive: circle; radius: 0.02;"
                    material="color: #FFFFFF; shader: flat;"
                    visible="false">
                </a-entity>
            </a-entity>
        </a-entity>

        <a-entity laser-controls="hand: left" raycaster="objects: .collidable; far:5000" line="color: #FFFFFF" class="laser-controls" visible="false"></a-entity>
        <a-entity laser-controls="hand: right" raycaster="objects: .collidable; far:5000" line="color: #FFFFFF" class="laser-controls" visible="false"></a-entity>

				<a-light type="ambient" color="#FFFFFF"></a-light>


        @if (isset($content['photo-spheres']) && count($content['photo-spheres']) > 0) 

        <a-entity 
						data-current-page="1" 
						isvr-photosphere-menu 
						id="photosphere-menu" 
						visible="false">
            @if (count($content['photo-spheres']) > 3)
            <a-entity isvr-photosphere-menu-navigation id="menu-arrow-up" position="0 1.5 0" visible="false" geometry="primitive: plane; width: 1; height: 0.5" material="transparent: true; opacity: 0">
                <a-plane position="-0.07 0 0" rotation="0 0 -45" width="0.10" height="0.3" color="#0080e5"></a-plane>
                <a-plane position="0.07 0 0" rotation="0 0 45" width="0.10" height="0.3" color="#0080e5"></a-plane>
            </a-entity>
            @endif
            @if (count($content['photo-spheres']) > 0) 
            <a-plane
								isvr-photosphere-menu-thumb="id: #photosphere-thumb-1" 
								class="collidable" 
								width="2" 
								height="0.7" 
								visible="false" 
								position="0 0.8 0.501"> 
            </a-plane>
            <a-image 
								data-image-id="1" 
								data-content-id="{{ $content['photo-spheres'][0]['photo-sphere']['#content-id'] }}" 
								id="photosphere-thumb-1" 
								class="img-photosphere-thumb" 
								width="2" 
								height="0.7" 
								position="0 0.8 0" 
								animation__mouseenter="property: position; from: 0 0.8 0; to: 0 0.8 0.5; dur: 400; startEvents: trigger-mouseenter" 
								animation__mouseleave="property: position; from: 0 0.8 0.5; to: 0 0.8 0; dur: 400; startEvents: trigger-mouseleave" 
								animation__fade="property: material.opacity; to: 1; startEvents: fade" 
								visible="false">
            </a-image>
            <a-ring 
								position="0 0.8 0.1" 
								color="#0080e5" 
								radius-inner="0.0625" 
								radius-outer="0.125" 
								theta-length="310" 
								animation__rotation="property: rotation; to: 0 0 -360; dur: 5000; easing: linear; loop: true; autoplay: true; startEvents: photosphere-loading-anim-1; pauseEvents: stop-photosphere-loading-anim-1" 
								id="photosphere-loading-1">
            </a-ring>
            @endif
            @if (count($content['photo-spheres']) > 1) 
            <a-plane
								isvr-photosphere-menu-thumb="id: #photosphere-thumb-2" 
								class="collidable" 
								width="2" 
								height="0.7" 
								visible="false" 
								position="0 0 0.501"> 
            </a-plane>
            <a-image 
								data-image-id="2" 
								data-content-id="{{ $content['photo-spheres'][1]['photo-sphere']['#content-id'] }}" 
								id="photosphere-thumb-2" 
								class="img-photosphere-thumb" 
								width="2" 
								height="0.7" 
								position="0 0 0" 
								animation__mouseenter="property: position; from: 0 0 0; to: 0 0 0.5; dur: 400; startEvents: trigger-mouseenter" 
								animation__mouseleave="property: position; from: 0 0 0.5; to: 0 0 0; dur: 400; startEvents: trigger-mouseleave" 
								animation__fade="property: material.opacity; to: 1; startEvents: fade" 
								visible="false">
            </a-image>
            <a-ring 
								position="0 0 0.1" 
								color="#0080e5" 
								radius-inner="0.0625" 
								radius-outer="0.125" 
								theta-length="310" 
								animation__rotation="property: rotation; to: 0 0 -360; dur: 5000; easing: linear; loop: true; autoplay: true; startEvents: photosphere-loading-anim-2; pauseEvents: stop-photosphere-loading-anim-2" 
								id="photosphere-loading-2">
            </a-ring>
            @endif
            @if (count($content['photo-spheres']) > 2) 
            <a-plane
								isvr-photosphere-menu-thumb="id: #photosphere-thumb-3" 
								class="collidable" 
								width="2" 
								height="0.7" 
								visible="false" 
								position="0 -0.8 0.501"> 
            </a-plane>
            <a-image 
								data-image-id="3" 
								data-content-id="{{ $content['photo-spheres'][2]['photo-sphere']['#content-id'] }}" 
								id="photosphere-thumb-3" 
								class="img-photosphere-thumb" 
								width="2" 
								height="0.7" 
								position="0 -0.8 0" 
								animation__mouseenter="property: position; from: 0 -0.8 0; to: 0 -0.8 0.5; dur: 400; startEvents: trigger-mouseenter" 
								animation__mouseleave="property: position; from: 0 -0.8 0.5; to: 0 -0.8 0; dur: 400; startEvents: trigger-mouseleave" 
								animation__fade="property: material.opacity; to: 1; startEvents: fade" 
								visible="false">
            </a-image>
            <a-ring 
								position="0 -0.8 0.1" 
								color="#0080e5" 
								radius-inner="0.0625" 
								radius-outer="0.125" 
								theta-length="310" 
								animation__rotation="property: rotation; to: 0 0 -360; dur: 5000; easing: linear; loop: true; autoplay: true; startEvents: photosphere-loading-anim-3; pauseEvents: stop-photosphere-loading-anim-3" 
								id="photosphere-loading-3">
            </a-ring>
            @endif
            @if (count($content['photo-spheres']) > 3) 
            <a-entity isvr-photosphere-menu-navigation="url:{{ $space_url }}/content/photo-spheres?per-page=3&page=2" id="menu-arrow-down" position="0 -1.5 0" visible="@if (count($content['photo-spheres']) > 3) true @else false @endif" geometry="primitive: plane; width: 1; height: 0.5" material="transparent: true; opacity: 0">
                <a-plane position="-0.07 0 0" rotation="0 0 45" width="0.10" height="0.3" color="#0080e5"></a-plane>
                <a-plane position="0.07 0 0" rotation="0 0 -45" width="0.10" height="0.3" color="#0080e5"></a-plane>
            </a-entity>
            @endif
        </a-entity><!-- photosphere menu //-->


        <a-sky
						class="collidable"
            isvr-init-assets="url:{{ $space_url }}/content/photo-spheres?per-page=3&page=1"
            animation__fadeout="property: material.color; from: #FFFFFF; to: #000000; dur: 500; startEvents: photosphere-fade-out"
            animation__fadein="property: material.color; from: #000000; to: #FFFFFF; dur: 500; startEvents: photosphere-fade-in"
						id="photosphere">
        </a-sky>


        @foreach ($content['photo-spheres'] as $photosphere)

            @if (trim($photosphere['title']['#value']) != '')

            		@include('theme::partials.photosphere_title_box_partial')

            @endif


            @if (isset($photosphere['attach-annotations']))
                @foreach ($photosphere['attach-annotations']['#positions'] as $annotation)

                    @php 
                    $rand = str_random();
                    @endphp

                    <a-entity 
                        rotation="{{ $annotation['#rotation']['#x'] }} {{ $annotation['#rotation']['#y'] }} {{ $annotation['#rotation']['#z'] }}">
                        <a-circle 
                            class="collidable hotspot hotspot-content-id-{{ $photosphere['attach-annotations']['#content-id'] }}" 
                            data-content-id="{{ $photosphere['attach-annotations']['#content-id'] }}" 
                            data-text-content-id="{{ $annotation['#content-id'] . $rand }}"
                            isvr-hotspot-wrapper-listener
                            material="transparent: false; opacity: 0"
                            position="0 1.6 -2.1" 
                            radius="0.4" 
                            scale="0.5 0.5 0.5" 
                            visible="false">
                            <a-circle 
                                color="{{ $annotation['#content']['background-color']['#value'] }}" 
                                material="transparent: false; opacity: 0.6"
                                radius="0.2" 
                                position="0 0 0.01"> 
                                <a-ring
                                    color="#FFFFFF"
                                    position="0 0 0.02"
                                    animation="property: geometry.radiusOuter; to: 0.15; dur: 1300; loop: true; easing: linear; dir: alternate"
                                    radius-inner="0.05"
                                    radius-outer="0.13">
                                    <!-- capture mouseover / mouseout events; enables smooth cursor animation //-->
                                    <a-circle
                                        material="opacity: 0"
                                        position="0 0 0.04"
                                        radius="0.4">
                                    </a-circle>
                                    <a-circle
                                        color="{{ $annotation['#content']['background-color']['#value'] }}"
                                        radius="0.05"
                                        position="0 0 0.03">
                                    </a-circle>
                                </a-ring>
                            </a-circle>
                        </a-circle>

                        @include('theme::partials.hotspot_text_box_partial')

                    </a-entity>

                @endforeach

            @endif

        @endforeach

        @endif


        <a-entity
            id="intro-0"
            class="collidable"
            visible="true"
            position="0 1.6 -2.1"
            geometry="primitive: plane; width: 1.8; height: 1"
            material="color: #FFFFFF; transparent: true; opacity: 0.5">
            <a-entity
                geometry="primitive: plane; width: 1.74; height: 0.94"
                position="0 0 0.01"
                material="color: #606060">
                <a-entity 
                    geometry="primitive: plane; width: 1.6; height: 0.4"
                    position="0 0.2 0.02" 
										material="shader: html; target: #space-title-texture; transparent: false; ratio: width">
                </a-entity>
                <a-ring 
                    position="0 -0.2 0.03" 
                    color="#000000" 
                    radius-inner="0.108" 
                    radius-outer="0.145" 
                    id="photosphere-loading-background">
                </a-ring>
                <a-ring 
                    position="0 -0.2 0.04" 
                    color="#0080e5" 
                    radius-inner="0.105" 
                    radius-outer="0.145" 
                    theta-length="120" 
                    animation="property: rotation; dur: 1000; to: 0 0 -360; easing: linear; loop: true; autoplay: true; pauseEvents: stop-photosphere-loading-anim" 
                    id="photosphere-loading">
                </a-ring>

                <a-entity 
                    id="photosphere-start-btn"
                    visible="false">
                    <a-circle
                    		id="photosphere-start-btn-circle"
                        position="0 -0.2 0.03" 
                        radius="0.125" 
                        animation="property: geometry.radius; to: 0.140; dur: 1300; dir: alternate; loop: true; easing: linear; pauseEvents: stop-photosphere-start-btn-anim"
                        color="#0080e5">
                    </a-circle>
                    <a-text
                        position="0.717 -0.194 0.04" 
                        value="Start"
                        color="#FFFFFF"
                        anchor="center"
                        width="1.6">
                    </a-text>
                    <!-- capture mouseover / mouseout events; enables smooth cursor animation //-->
                    <a-entity
                        material="opacity: 0"
                        geometry="primitive: plane; width: 1.5; height: 0.7"
                        position="0 0 0.09">
                    </a-entity>
                </a-entity>
            </a-entity>
        </a-entity><!-- intro-0 //-->

    </a-scene>


		<div class="cover">
    </div>


    <div id="space-title-texture" style="color:#FFFFFF;background-color:#606060">
    {!! $space_title !!}
    </div>


		@if (isset($content['photo-spheres']))
        @foreach ($content['photo-spheres'] as $photosphere)
            @if (trim($photosphere['title']['#value']) != '')
                <div id="photosphere-title-texture-content-id-{{ $photosphere['photo-sphere']['#content-id'] }}" class="photosphere-title-texture" style="background-color:{{ $photosphere['background-color']['#value'] }}; color:{{ $photosphere['text-color']['#value'] }}">
                {!! $photosphere['title']['#value'] !!}
                </div>
            @endif
        @endforeach
    @endif


    @if (isset($content['annotations']))
        @foreach ($content['annotations'] as $annotation)
            <div id="annotation-text-texture-content-id-{{ $annotation['text']['#content-id'] }}" class="annotation-text-texture" style="background-color:{{ $annotation['background-color']['#value'] }}; color:{{ $annotation['text-color']['#value'] }}">
            {!! nl2br($annotation['text']['#value']) !!}
            </div>
        @endforeach
    @endif

@endsection


